### PR TITLE
Ensure synchronization when taking snapshots

### PR DESCRIPTION
**Pull request by pirgeo**
Currently, it is possible that data is written to the Dynatrace Summary object during the creation of a snapshot. An example: the sum data is read (`totalAmount` is called), then a different thread adds an observation to the underlying summary, and only then `count()` is called to retrieve the number of elements. By setting the monitor on the DynatraceSummary object using the synchronized keyword, the creation of the Snapshot becomes an atomic operation. The write method locks the same object (https://github.com/dynatrace-oss-contrib/micrometer/blob/main/implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/types/DynatraceSummary.java#L43). So, while the snapshot is created and all the reads are processed, no new data can be added. Similarly, while data is added, no snapshot can be taken.

### CODE REVIEW

**Comment by pirgeo** [comment]
Needs some more work, will reopen once it's ready.

**Comment by pirgeo** [comment]
Reopened the PR. In the updated version, I moved the synchronization logic for the taking of a snapshot to the innermost place (in the `DynatraceSummary`). This way, it should never be possible for code relying on the `DynatraceSummary` to create a Snapshot that is not synchronized.

For readability, I also split out the `DynatraceSummarySnapshotSupport`. There is a new Interface, `TimeAwareDynatraceSummarySnapshotSupport` that contains the Snapshot functionality for Timers. That way, the Timer does not have to implement the unitless functionality, and the DistributionSummary does not have to implement the one with units.

File: implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/v2/DynatraceExporterV2.java [hunk 1]
```diff
@@ -198,18 +199,17 @@ private String createCounterLine(Meter meter, Measurement measurement) {
     }
 
     Stream<String> toTimerLine(Timer meter) {
-        if (!(meter instanceof DynatraceSummarySnapshotSupport)) {
-            // fall back to previous behaviour
+        if (!(meter instanceof TimeAwareDynatraceSummarySnapshotSupport)) {
             return toSummaryLine(meter, meter.takeSnapshot(), getBaseTimeUnit());
```
Comment (joaopgrassi) [inline implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/v2/DynatraceExporterV2.java:203]: Here we call `getBaseTimeUnit()`, but inside `meter.takeSnapshot()`, (DynatraceTimer:106) we do `baseTimeUnit()`. I'm a bit confusing how those differ? The former seems to come from our config but the other from MM? 🤔

File: implementations/micrometer-registry-dynatrace/src/test/java/io/micrometer/dynatrace/types/DynatraceTimerTest.java [hunk 2]
```diff
@@ -50,24 +50,30 @@ class DynatraceTimerTest {
...
+        assertThat(timer.count()).isEqualTo(2);
 
-        // checks that the recorded values are returned in the TimeUnit used to set up the
-        // instrument
-        assertMinMaxSumCount(timer, 314, 476, 790, 2);
-        assertThat(timer.hasValues()).isTrue();
-        assertMinMaxSumCount(timer.takeSummarySnapshotAndReset(), 314, 476, 790, 2);
-        assertThat(timer.hasValues()).isFalse();
+        timer.takeSummarySnapshotAndReset(BASE_TIME_UNIT);
+        assertThat(timer.count()).isZero();
```
Comment (joaopgrassi) [inline implementations/micrometer-registry-dynatrace/src/test/java/io/micrometer/dynatrace/types/DynatraceTimerTest.java:65]: Any reason you split asserting the count + the recorded values? I was confused for a bit why we stopped asserting only to see the test below. 

Also empty line here.

**Review by joaopgrassi — APPROVED** [review]
Just some nits and question. Not blocking.

**Review by pichlermarc — APPROVED** [review]
My Micrometer knowledge is a bit limited, but looks good to me :slightly_smiling_face:

**Comment by pirgeo** [comment]
I added some tests for unit conversions. There were some (https://github.com/dynatrace-oss-contrib/micrometer/blob/main/implementations/micrometer-registry-dynatrace/src/test/java/io/micrometer/dynatrace/types/DynatraceTimerTest.java#L149) but more is better in this case, I think.

**Review by pichlermarc — APPROVED** [review]
Thanks for adding the tests :slightly_smiling_face:

**Pull request merged**